Stop concat from attempting to sort mismatched columns by default

brycepg opened the pull request:
Preserve column order upon concatenation to obey
least astonishment principle.

Allow old behavior to be enabled by adding a boolean switch to
concat and DataFrame.append, mismatch_sort, which is by default disabled.  

- [x] Close #4588
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

brycepg committed:
Stop concat from attempting to sort mismatched columns by default

Preserve column order upon concatenation to obey
least astonishment principle.

Allow old behavior to be enabled by adding a boolean switch to
concat and DataFrame.append, mismatch_sort, which is by default disabled.

Close #4588

brycepg committed:
Stop concat from attempting to sort mismatched columns by default

Preserve column order upon concatenation to obey
least astonishment principle.

Allow old behavior to be enabled by adding a boolean switch to
concat and DataFrame.append, mismatch_sort, which is by default disabled.

Close #4588

pep8speaks commented:
Hello @brycepg! Thanks for submitting the PR.

 - In the file [`pandas/core/frame.py`](https://github.com/pandas-dev/pandas/blob/db4c521e195e7e8f385c65474d54fbfb2b178218/pandas/core/frame.py), following are the PEP8 issues :

> [Line 5985:80](https://github.com/pandas-dev/pandas/blob/db4c521e195e7e8f385c65474d54fbfb2b178218/pandas/core/frame.py#L5985): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 79 characters)

 - In the file [`pandas/core/indexes/api.py`](https://github.com/pandas-dev/pandas/blob/db4c521e195e7e8f385c65474d54fbfb2b178218/pandas/core/indexes/api.py), following are the PEP8 issues :

> [Line 77:80](https://github.com/pandas-dev/pandas/blob/db4c521e195e7e8f385c65474d54fbfb2b178218/pandas/core/indexes/api.py#L77): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)



 - In the file [`pandas/tests/reshape/test_concat.py`](https://github.com/pandas-dev/pandas/blob/db4c521e195e7e8f385c65474d54fbfb2b178218/pandas/tests/reshape/test_concat.py), following are the PEP8 issues :

> [Line 855:80](https://github.com/pandas-dev/pandas/blob/db4c521e195e7e8f385c65474d54fbfb2b178218/pandas/tests/reshape/test_concat.py#L855): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 1014:80](https://github.com/pandas-dev/pandas/blob/db4c521e195e7e8f385c65474d54fbfb2b178218/pandas/tests/reshape/test_concat.py#L1014): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 2162:52](https://github.com/pandas-dev/pandas/blob/db4c521e195e7e8f385c65474d54fbfb2b178218/pandas/tests/reshape/test_concat.py#L2162): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 2163:52](https://github.com/pandas-dev/pandas/blob/db4c521e195e7e8f385c65474d54fbfb2b178218/pandas/tests/reshape/test_concat.py#L2163): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

jreback commented on pandas/core/reshape/concat.py:
```diff
@@ -20,7 +20,7 @@
 def concat(objs, axis=0, join='outer', join_axes=None, ignore_index=False,
            keys=None, levels=None, names=None, verify_integrity=False,
-           copy=True):
+           copy=True, mismatch_sort=False):
```
just call this sort 
put before copy

jreback reviewed (changes requested):
needs a few more tests
replicate any tests from the original issue
needs a whatsnew note
will comment on the impl later

jorisvandenbossche commented:
Thanks for working on this!

* We also need to check how this works in combination with other keywords (eg `join`)
* Do we want to ignore the keyword if all columns match? (what you currently do?)  I would also find that surprising that the keyword then does not work. And given that `sort=False` is the default, I think we can let `sort=True` sort always?

brycepg commented:
@jreback 

I've make the requested changes

@jorisvandenbossche 

* Yeah it does look like `concat` is used a lot internally. I'll have a look at each location its called.
* That sounds reasonable

jorisvandenbossche commented:
> Yeah it does look like concat is used a lot internally. I'll have a look at each location its called.

Ah, yes, that might be needed as well. But what I meant was other keywords of concat itself. Eg does the sort keyword work for both `join='inner'` and `join='outer'` ?

TomAugspurger commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -1160,6 +1160,7 @@ Reshaping
 - Bug in :meth:`DataFrame.astype` where column metadata is lost when converting to categorical or a dictionary of dtypes (:issue:`19920`)
 - Bug in :func:`cut` and :func:`qcut` where timezone information was dropped (:issue:`19872`)
 - Bug in :class:`Series` constructor with a ``dtype=str``, previously raised in some cases (:issue:`19853`)
+- Stop :func:`concat` and ``Dataframe.append`` from sorting columns by default. Use ``sort=True`` to retain old behavior (:issue:`4588`)
```
This should be ``sort=True`` by default to preserve backwards compatibility, right?

Or rather, I think the eventual goal is to have ``sort=False`` be the default, so for now it should be

- ``sort=None`` is the default
- If the default is passed, use ``sort=True`` and warn that the default is changing in the futrue
- If True or False, no warning.

jreback replied:
actually this needs a sub-section. this is a rather large change (even if its None by default). highliting it is best. pls show an example of previous and new

brycepg replied:
@TomAugspurger
If I do what @jorisvandenbossche suggests then, `sort=True` will not be backwards compatible because it will sort the axes in question regardless of whether the columns are mismatched.
I could have `sort=None` be the default, give a warning and revert to old behavior. In future versions this behavior of only sorting the axes sometimes would not be available because it doesn't make sense and concat could default to `sort=False`

jorisvandenbossche replied:
@brycepg I think we can do both (it might complicate the code a bit, but not too much I think, as in `_get_combined_index` those cases are already handled separately). As @TomAugspurger suggests, the default can be None for now, so we can raise a warning in the appropriate cases:

- when all axes are equal (current case when there is already no sorting): no warning should be raised, as the future default of `sort=False` will not change anything, but add the ability to also sort the index with `sort=True`
- when not all axes are equal (current case that the unwanted sorting happens): issue a warning that in the future this will no longer sort

jreback commented on pandas/core/frame.py:
```diff
@@ -5982,7 +5982,8 @@ def infer(x):
 
-    def append(self, other, ignore_index=False, verify_integrity=False):
+    def append(self, other, ignore_index=False,
+               verify_integrity=False, sort=False):
```
sort before verify_integrity

TomAugspurger replied:
@jreback why do you want sort before verify_integrity?

jreback commented on pandas/core/frame.py:
```diff
@@ -5995,6 +5996,8 @@ def append(self, other, ignore_index=False, verify_integrity=False):
         verify_integrity : boolean, default False
             If True, raise ValueError on creating index with duplicates.
+        sort: boolean, default False
+            Sort columns if given object doesn't have the same columns
```
needs a versionadded.

use does not

jreback commented on pandas/core/reshape/concat.py:
```diff
@@ -20,7 +20,7 @@
 
 def concat(objs, axis=0, join='outer', join_axes=None, ignore_index=False,
            keys=None, levels=None, names=None, verify_integrity=False,
-           copy=True):
```
actually move before verify_integrity

jreback replied:
pls do this

TomAugspurger replied:
Why make this an API breaking change?

jreback replied:
because it more logical.

TomAugspurger replied:
How so?

I'm OK with breaking API when necessary, but this seems unnecessary.

jreback commented on pandas/core/reshape/concat.py:
```diff
@@ -60,6 +60,8 @@ def concat(objs, axis=0, join='outer', join_axes=None, ignore_index=False,
     verify_integrity : boolean, default False
         Check whether the new concatenated axis contains duplicates. This can
         be very expensive relative to the actual data concatenation
+    sort : boolean, default False
```
simiar to above

jreback commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -2155,3 +2155,24 @@ def test_concat_empty_and_non_empty_series_regression():
+    dfa = pd.DataFrame(columns=['C', 'A'], data=[[1, 2]])
+    dfb = pd.DataFrame(columns=['C', 'Z'], data=[[5, 6]])
+    result = pd.concat([dfa, dfb])
+    assert result.columns.tolist() == ['C', 'A', 'Z']
```
create an expected frame and use assert_frame_equal

jreback commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -2155,3 +2155,24 @@ def test_concat_empty_and_non_empty_series_regression():
+    df['a'] = [1, 2, 3]
+    df2 = pd.DataFrame({'a': [4, 5]})
+    df3 = pd.concat([df, df2])
+    assert df3.columns.tolist() == ['b', 'c', 'a']
```
same

jreback commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -2155,3 +2155,24 @@ def test_concat_empty_and_non_empty_series_regression():
+    df['c'] = [1, 2, 3]
+    df['a'] = [1, 2, 3]
+    df2 = pd.DataFrame({'a': [4, 5]})
+    df3 = pd.concat([df, df2])
```
use result =

brycepg commented:
Sure I’ll try to do it tonight

On Mon, Apr 23, 2018 at 1:15 PM, Tom Augspurger <notifications@github.com> wrote:

> [@brycepg](https://github.com/brycepg) can you update based on Jeff's comments? Doing a release candidate soon (tomorrow or Wednesday hopefully), and it'd be nice to have this in.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, [view it on GitHub](https://github.com/pandas-dev/pandas/pull/20613#issuecomment-383690194), or [mute the thread](https://github.com/notifications/unsubscribe-auth/AAa9yC_mpbFjUbWvFWVaeM4sWVQ3pxffks5trig-gaJpZM4THzks).
>
> {"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/pandas-dev/pandas","title":"pandas-dev/pandas","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/pandas-dev/pandas"}},"updates":{"snippets":[{"icon":"PERSON","message":"@TomAugspurger in #20613: @brycepg can you update based on Jeff's comments? Doing a release candidate soon (tomorrow or Wednesday hopefully), and it'd be nice to have this in."}],"action":{"name":"View Pull Request","url":"https://github.com/pandas-dev/pandas/pull/20613#issuecomment-383690194"}}}

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into brycepg-master

TomAugspurger committed:
Updates

API: Updated the default to be compatible and warn.

DOC: updated the whatsnew and concat docstring.

TomAugspurger committed:
Updates

API: Updated the default to be compatible and warn.

DOC: updated the whatsnew and concat docstring.

TomAugspurger commented:
@brycepg I pushed some changes to your branch.

1. I updated the API to keep this backwards compat, but you'll get a warning.
2. I updated the docs to explain that

Will you have time today to get go through and update the tests to pass `sort=True` where necessary? The build log will have a bunch of warnings when it finishes.

And I didn't address most of @jreback's comments yet.

I'd like to do a release candidate tomorrow, so if you can't get to it today let me know and I'll push more fixes here.

TomAugspurger committed:
Test fallout

TomAugspurger committed:
Test fallout

TomAugspurger committed:
Updated append

TomAugspurger committed:
Updated append

TomAugspurger committed:
versionadded

TomAugspurger committed:
versionadded

TomAugspurger commented:
Can someone proofread the warning text? Specifically, does the term "non-concatenation axis" make sense? I could also call it the "non-expanding axis", or something else entirely.

TomAugspurger committed:
Squash more test warnings

TomAugspurger committed:
Squash more test warnings

TomAugspurger committed:
py2 compat

TomAugspurger committed:
py2 compat

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into brycepg-master

TomAugspurger commented:
Sorry, that was a bad example. join='inner' already preserves the column order of the intersection

```python

In [4]:     df1 = pd.DataFrame({"a": [1, 2], "b": [1, 2], "c": [1, 2]},
   ...:                        columns=['b', 'a', 'c'])
   ...:     df2 = pd.DataFrame({"a": [1, 2], 'b': [3, 4]}, index=[3, 4])
   ...:     result = pd.concat([df1, df2], join='inner')
   ...:
   ...:     result
   ...:
   ...:
Out[4]:
   b  a
0  1  1
1  2  2
3  3  1
4  4  2
```

I'll document that `sort` has no effect for `join='inner'`.

TomAugspurger committed:
Document outer is not affected

TomAugspurger committed:
Document outer is not affected

TomAugspurger committed:
Docs

TomAugspurger committed:
Docs

TomAugspurger commented:
> Do we want to ignore the keyword if all columns match?

This depends a bit on the eventual fate of the `sort` keyword. While we may not ever remove it, since we don't want to introduce a keyword and deprecate it shortly after, we can at least discourage its use once the default has been changed. I could imagine the docstring for `sort` eventually just being "for backwards compatibility. Has no effect."

If people want to sort, they should sort that axis and then reindex before or after the concat.

jorisvandenbossche commented:
> This depends a bit on the eventual fate of the sort keyword.

What's the problem with keeping the keyword, and let it mean to always sort? (sorry if this makes no sense, it's a while ago I looked at this)

TomAugspurger commented:
No problem, just not useful I think.

Though passing sort=True and not having it sort, for already aligned, would be strange. I’ll push an update for that later today.

jreback commented on doc/source/merging.rst:
```diff
@@ -181,8 +181,14 @@ behavior:
 
-Note that the row indexes have been unioned and sorted. Here is the same thing
-with ``join='inner'``:
+.. versionchanged:: 0.23.0
```
make this a note (maybe a warning)

jreback commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -639,6 +639,34 @@ Returning a ``Series`` allows one to control the exact return structure and colu
 
     df.apply(lambda x: Series([1, 2, 3], index=['D', 'E', 'F']), axis=1)
 
```
need a ref here

jreback commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -639,6 +639,34 @@ Returning a ``Series`` allows one to control the exact return structure and colu
+
+   pd.concat([df1, df2], sort=False)
+
+Note that this change also applies to :meth:`DataFrame.append`, which has also received a `sort` keyword for controlling this behavior.
```
use double backticks around sort

TomAugspurger replied:
Single backticks for parameter names.

jreback replied:
that doesn't render as nicely, and its inconsistent with the docs themselves. Unless you want to change all of them I would use double-backticks.

jorisvandenbossche replied:
Single backticks is the "rule" according to our docstring guidelines, but I agree we probably don't do it consistently following that rule in the narrative docs.

jreback replied:
this is not a doc-string

jreback commented on pandas/core/indexes/api.py:
```diff
@@ -89,13 +103,19 @@ def conv(i):
         index = indexes[0]
         for other in indexes[1:]:
             if not index.equals(other):
+
```
isn't this already passed in as ``sort=True``?

TomAugspurger replied:
What do you mean?

jreback replied:
I mean this is deeply buried here, why is this showing a user facing warning? IOW what is the user call that triggers this?

TomAugspurger replied:
> IOW what is the user call that triggers this?

`pd.concat` or `DataFrame.append` with unaligned objects and an unspecified `sort`.

Where else would it go? Before this point we don't know what the indexes to align are, and whether or not they're aligned.

jreback reviewed (changes requested):
would prefer that you add a sort fixture to the top-level conftest and test most things you changed to sort=True to use that fixture. a bit of work, but often this could surface some bugs. Also are there any remaining warnings?

TomAugspurger committed:
Sort for intersection

TomAugspurger committed:
Sort for intersection

TomAugspurger committed:
More tests

TomAugspurger committed:
More tests

TomAugspurger reviewed (commented):
One subtly that's probably unimportant. If you have an Index with incomparable objects (mix of ints and strs) then we'll silently fail to sort. That's consistent with how union works. I think it's uncommon enough that it's not a big deal.

TomAugspurger committed:
Test fixup.

Sparse as well

TomAugspurger committed:
Test fixup.

Sparse as well

TomAugspurger committed:
ugh

jreback commented on pandas/core/frame.py:
```diff
@@ -6051,6 +6052,14 @@ def append(self, other, ignore_index=False, verify_integrity=False):
             If True, do not use the index labels.
         verify_integrity : boolean, default False
             If True, raise ValueError on creating index with duplicates.
+        sort : boolean, default None
```
put before verify_integrity

jreback commented on pandas/core/indexes/api.py:
```diff
@@ -31,33 +44,40 @@
+        index = _union_indexes(indexes, sort=sort)
+        index = _ensure_index(index)
+
+    if sort and not index.is_monotonic_increasing:
```
this check is not needed (that's the first thing sort checks)

jreback commented on pandas/core/reshape/concat.py:
```diff
@@ -60,6 +60,19 @@ def concat(objs, axis=0, join='outer', join_axes=None, ignore_index=False,
     verify_integrity : boolean, default False
         Check whether the new concatenated axis contains duplicates. This can
         be very expensive relative to the actual data concatenation
+    sort : boolean, default None
```
same

jreback commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -793,7 +800,32 @@ def test_append_records(self):
+                                columns=['a', 'b', 'c'])
+        tm.assert_frame_equal(result, expected)
+
+        # sort=True, the previous behavior
```
you could just use the fixture and just change the expected if sort is False (like you do elsewhere)

jreback commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -2249,3 +2295,109 @@ def test_concat_empty_and_non_empty_series_regression():
+                             'c': [1, 2, 1, 2]},
+                            columns=['a', 'b', 'c'])
+    tm.assert_frame_equal(result, expected)
+
```
not that I don't mind more tests, but are any of these duplicates of existing ones?

TomAugspurger replied:
Yeah, probably. I'll try to de-duplicate a bit.

jorisvandenbossche commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -639,6 +639,36 @@ Returning a ``Series`` allows one to control the exact return structure and colu
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In a future version of pandas :func:`pandas.concat` will no longer sort the non-concatenation axis when it is not already aligned.
+The current behavior is the same as the previous (sorting), but now a warning is issued (:issue:`4588`).
```
I would mention the warning is only issues if the labels are not exactly the same.

TomAugspurger committed:
quoting

TomAugspurger committed:
Clarify

TomAugspurger committed:
Removed unnescesary check

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into brycepg-master

TomAugspurger committed:
Merge remote-tracking branch 'brycepg/master' into brycepg-master

TomAugspurger committed:
Prune tests

TomAugspurger committed:
Default sort

TomAugspurger committed:
Make both tests happy

TomAugspurger committed:
Explicit columns

TomAugspurger committed:
List of series

TomAugspurger commented:
Found one more issue in crosstab. Fixing now.

jreback commented on pandas/core/base.py:
```diff
@@ -507,7 +507,7 @@ def is_any_frame():
                            for r in compat.itervalues(result))
 
             if isinstance(result, list):
-                return concat(result, keys=keys, axis=1), True
```
thre are a bunch more concats in this same section. basically they control the resulting ordeing of the aggregation. I guess ``sort=True`` is fine here (for the other cases).

jreback replied:
or maybe these should be ``sort=False``, not sure what is actually affected (as these might be aligned ops already)

jreback commented on pandas/core/groupby/groupby.py:
```diff
@@ -1098,7 +1098,8 @@ def reset_identity(values):
 
                 result = concat(values, axis=self.axis, keys=group_keys,
-                                levels=group_levels, names=group_names)
+                                levels=group_levels, names=group_names,
```
same here, there are like 10 calls to concat. I think should be explicit about sort

jreback commented on pandas/core/indexes/api.py:
```diff
@@ -89,13 +110,19 @@ def conv(i):
         for other in indexes[1:]:
             if not index.equals(other):
+
+                if sort is None:
```
I would move ``_unique_indices`` from a nested function to module level (e.g. same as ``_union_indices`` (and maybe conform the spelling indices / indexes), and simply add a ``sort=`` kwarg (which you are already passing into fast_unique_multiple_lists). Then you can do the warning there. just makes this whole function a bit simpler.

jreback commented on pandas/core/panel.py:
```diff
@@ -1499,8 +1499,11 @@ def _extract_axis(self, data, axis=0, intersect=False):
+            # 1. we're doing a union (intersect=False)
+            # 2. the indices are not aligned.
             index = _get_objs_combined_axis(data.values(), axis=axis,
-                                            intersect=intersect)
```
so this always warns? shouldn't this be ``sort=False``?

TomAugspurger committed:
test, fix pivot

TomAugspurger commented:
K, I'll try to go through all the internal calls to concat and set explicitly.

One thing that's made this difficult is that we don't really have the old option of *sort only if not aligned* anymore. So just setting `sort=True` or `sort=False` won't necessarily reproduce the old output. Maybe all of these will be OK though.

TomAugspurger commented:
I'm going to push this to 0.23.1. I'm not confident enough that we have complete test coverage for all the cases of concat used internally.

I can pick it up after PyCon.

TomAugspurger commented:
Yeah. The concern is about internal calls to concat.

For example, `Panel({b: df, a: df})` was just broken on this branch for a
while with certain conditions. None of the constructor tests caught that,
it was just in a pytables test that it was caught. Though perhaps panels
isn't the best example, not sure how coverage is on that in general.

On Tue, May 1, 2018 at 2:24 PM, Joris Van den Bossche <
notifications@github.com> wrote:

> If we don't include it in 0.23, I think it has to wait for 0.24. We
> shouldn't introduce new deprecation warnings in a bug fix release.
>
> However, I don't fully understand the problem to not merge it now. There
> are still some internal use cases that you need to check? (whether
> sort=True/False needs to be added?)
> But as far as I see, there are no failing tests? Are there still warnings
> being raised?
> I think in many cases in internal code, it's concatting aligned results
> (eg in groupby), and then there should be no change or warning.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/20613#issuecomment-385763928>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIuROG_dJhxtzOuPjLcHphSagFb0Pks5tuLZSgaJpZM4THzks>
> .
>

jorisvandenbossche commented:
In what way was it broken? (there should for now mainly be a warning and new keyword, but not really change in default behaviour?)

>  I'm not confident enough that we have complete test coverage for all the cases of concat used internally.

I can't really say anything founded, as I didn't look enough into detail to what tests have broken throughout implementing this and what you needed to add, but since our tests are now passing, my gut feeling would say: let's use the rc period to get more real-world testing on that .. ;)

TomAugspurger commented:
> In what way was it broken? (there should for now mainly be a warning and new keyword, but not really change in default behaviour?)

I don't recall the exact circumstances, but a MultiIndex was (or wasn't?) being sorting.

If you're OK with merging as is, then I can commit to adding more tests (and fixing bugs) between now and Friday. I just don't think it should hold up the release.

jorisvandenbossche commented:
As I said, difficult to say myself, but I am OK with relying on your judgement here. 
(and I certainly approve the gist of the PR)

TomAugspurger commented:
Alright, let's do it.

https://github.com/pandas-dev/pandas/issues/20909 for the followup.
